vmgen: fix double evaluation of `swap` arguments

zerbina opened the pull request:
## Summary

Fix a bug the VM code generator had where the expression used
as the first argument was always evaluated twice -- it is now
only evaluated once.

## Details

The `lowerSwap` procedure that was used for lowering the `swap`
call into assignments duplicated the first argument expression,
resulting in it being evaluated two times.

Usage of `lowerSwap` is replaced with manually emitting the
bytecode for the assignments, which in addition to fixing the
bug and making the generated code slightly more efficient (the
introduced temporary is freed immediately after the swap now),
also has the benefit of removing both the import dependency on
`lowering` and `vmgen` needing access to an `IdGenerator`.

`lowerSwap` has no further usage sites, and is thus removed.

saem commented on compiler/vm/vmgen.nim:
```diff
@@ -1810,7 +1807,37 @@ proc genMagic(c: var TCtx; n: PNode; dest: var TDest; m: TMagic) =
+    # var types
+    if fitsRegister(n[1].typ):
+      # we need to account for the fact that either operand could be
+      # stored in a register alreay (because it's a local variable)
```
```suggestion
      # stored in a register already (because it's a local variable)
```

zerbina committed:
tests: clean up the `tswap.nim` test

* reference the issues via a full link
* replace echo with `doAssert`
* use the VM target

zerbina committed:
vmgen: implement `swap` without the lowering

zerbina committed:
tests: add tests

One for the fixed bug, the other for having explicit coverage of the
"operands of primitive type" case.

zerbina committed:
lowerings: remove `lowerSwap`

It has no usage sites left, and, given its defect, shouldn't be used
again in in the future.

saem reviewed (approved):
minor typo; still readable.

zerbina committed:
fix a typo

Co-authored-by: Saem Ghani <saemghani+github@gmail.com>

saem commented:
/merge
